### PR TITLE
Sockets - Erica & Karla G.

### PR DESCRIPTION
**Pull request by kguadron**
# Adagrams
Congratulations! You're submitting your assignment.

## Comprehension Questions

Feature | Feedback
--- | ---
What are the components that make up a method? | A method is made up of the method signature and the method block.
What are the advantages of using git when collaboratively working on one code base? | The advantage is that it's easy to see what other collaborators are doing and share/merge changes.
What kind of relationship did you and your pair have with the unit tests? | It was helpful for debugging and seeing where in the code the errors were coming up.
Does your code use any methods from the `Enumerable` mixin? If so, where and why was it helpful? | We used .each and .each_char. It was helpful in iterating through the individual elements in the different arrays.
What was one method you and your pair used to debug code? | We used the unit tests and puts statements.
What are two discussion points that you and your pair discussed when giving/receiving feedback from each other that you would be willing to share? | 1) We have different coding paces and acknowledged that it's important to strike a balance. 2) We fostered a safe/comfortable working dynamic where we were able to communicate needs, ideas, and points of confusion.

### CODE REVIEW

File: lib/adagrams.rb [hunk 1]
```diff
@@ -0,0 +1,103 @@
+require "csv"
+
+def draw_letters
+  available_letters = {"a" => 9, "b" => 2, "c" => 2, "d" => 4, "e" => 12, "f" => 2, "g" => 3, "h" => 2,
+                       "i" => 9, "j" => 1, "k" => 1, "l" => 4, "m" => 2, "n" => 6, "o" => 8, "p" => 2,
+                       "q" => 1, "r" => 6, "s" => 4, "t" => 6, "u" => 4, "v" => 2, "w" => 2, "x" => 1,
```
Comment (droberts-sea) [inline lib/adagrams.rb:6]: I love this data structure - it makes it very clear how many there are of each letter. Could you store it in a constant?

File: lib/adagrams.rb [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+                       "y" => 2, "z" => 1}
+  drawn_letters = []
+  10.times do
+    letter = available_letters.keys.sample
+    while available_letters[letter] == 0
+      letter = available_letters.keys.sample
```
Comment (droberts-sea) [inline lib/adagrams.rb:12]: By this logic, you're as likely to draw an `e` as you are a `z`!

Instead, you might create an array containing all the letters, with code similar to this:

```ruby
letter_counts = {
  # the hash you had before
}
letters = []
letter_counts.each do |letter, count|
  count.times do
    letters << letter
  end
end
```

Then you could pull a random tile with `letters.sample` and get a nice even distribution.

File: lib/adagrams.rb [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+    available_letters[letter] = available_letters[letter] - 1
+  end
+  return drawn_letters
+end
+
+# Method that deletes element in array at its specific index. Prevents deletion of mathcing elements
+class Array
+  def delete_elements_in_(array)
+    array.each do |x|
+      if index = index(x)
```
Comment (droberts-sea) [inline lib/adagrams.rb:24]: Be careful about this sort of monkey patching - it's typically considered bad form. Instead, you could write a helper method that's _not_ part of the `Array` class:

```ruby
def delete_from(hay, needles)
  needles.each do |needle|
    if index = hay.index(needle)
      hay.delete_at(index)
    end
  end
end
```

File: lib/adagrams.rb [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+  end
+end
+
+# Checks if user inputs letters from their letter bank
+def uses_available_letters?(input, letters_in_hand)
+  used_letters = []
+  input.each_char do |letter|
+    if letters_in_hand.include?(letter) == true
+      letters_in_hand.delete_elements_in_([letter])
+      used_letters << letter
```
Comment (droberts-sea) [inline lib/adagrams.rb:37]: Since you only use `delete_elements_in_` with an array of length 1, it might make sense to make the argument be a single object instead of an array.

Please make sure you fully understand the code you borrow from other sources before including it in your project.

File: lib/adagrams.rb [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+#Wave 3
+def score_word(word)
+  word = word.downcase
+  score = 0
+  word.each_char do |letter|
+    case letter
+    when "a", "e", "i", "o", "u", "l", "n", "r", "s", "t"
+      score += 1
+    when "d", "g"
+      score += 2
```
Comment (droberts-sea) [inline lib/adagrams.rb:56]: While this `case` statement works, it means that the information about which letter has which score is locked into this piece of code, and can't easily be used elsewhere. For example, if you wanted to display the value of each letter in a hand, you would need to repeat this work.

An alternative approach would be to store the letter scores in a hash, something like this:

```ruby
LETTER_SCORES = {
  "A" => 1
  "B" => 3,
  "C" => 3,
  "D" => 2,
  # ...
}
```

Then to get the score for a letter, you can say `LETTER_SCORES[letter]`.

**Comment by droberts-sea** [comment]
# Adagrams
## What We're Looking For

Feature | Feedback
--- | ---
**General** |
Answered comprehension questions | yes
Small commits with meaningful commit messages | yes - good work!
**Code Requirements** |
`draw_letters` method |
Uses appropriate data structure to store the letter distribution | see inline
All tests for `draw_letters` pass | yes
`uses_available_letters?` method |
All tests for `uses_available_letters?` pass | yes
`score_word` method |
Uses appropriate data structure to store the letter scores | see inline
All tests for `score_word` pass | yes
`highest_score_from` method | 
Appropriately handles edge cases for tie-breaking logic | yes
All tests for `highest_score_from` pass | yes
**Overall** |



Good job overall. There are a few places where things could be cleaned up, which I've tried to call out below, but in general I am happy with what I see here. It's clear to me that the learning goals around working with methods and tests were met. Keep up the hard work!
